Fixed the problem with escorts (I think).

onpon4 opened the pull request:
Not tested thoroughly yet, but I think this should fix the problem of
escorts causing segfaults when their origin ship dies. A cursory test
shows that this seems to work.

This also causes pilot.setLeader to remove "deployed" status from the
origin ship.

ids1024 commented on src/pilot.c:
```diff
@@ -2903,6 +2909,14 @@ void pilot_destroy(Pilot* p)
    memmove(&pilot_stack[i], &pilot_stack[i+1], (pilot_nstack-i)*sizeof(Pilot*));
+
+   /* Clear docks for launched fighters if necessary */
+   for (i=0; i < pilot_nstack; i++) {
```
I think this will work...

But it means every `pilot_destroy` call iterates over every pilot in the system, which might not be the most efficient solution.

If instead `dockpilot` is a pilot ID, and `dockslot` an index in the slots, I think that would be both more efficient and easier not to break (since there aren't pointers lying around to potentially become dangling).

ids1024 replied:
I think using a pilot ID and slot index should work nicely with a helper function like this (untested):

```c
PilotOutfitSlot* pilot_getDockSlot( Pilot* p ) {
   Pilot* dockpilot;
   if (p->dockpilot != 0) {
      docpilot = pilot_get(p->dockpilot);
      if (docpilot == NULL)
         p->docpilot = 0;
      else
         return docpilot->outfits[p->dockslot];
   }
   return NULL;
}
```

onpon4 replied:
That's real fair. It did occur to me that would be cleaner as an implementation, too.

onpon4 replied:
OK so I'm finding implementing it that way to be a bit cumbersome. Giving up on that for now, I don't have the energy for it.

(Also worth noting, that change isn't going to reduce looping at all, the looping is to find the ships that mark the pilot being destroyed as a dock and changing the way said pilot is marked won't prevent having to loop to do that. The only way to avoid that is if we use a list stored for each pilot that shows all the pilots that use them as a dock, which I guess the escort list *might* work for, but I haven't looked into that yet.)

ids1024 replied:
> that change isn't going to reduce looping at all

With the way I suggested, I *think* this loop could be removed. When the escort's code tries to call `pilot_getDockSlot()`, that will simply return `NULL` since the pilot that launched it wouldn't exist any more; without `docpilot`/`docslot` having to be updated already.

onpon4 replied:
Alright, as per usual after whining for a bit I did the conversion. It's using ints for both dockpilot and dockslot now.

ids1024 commented:
Calling `assert` broke the Travis build. If you want to do that, include `assert.h`.

ids1024 commented on src/player.h:
```diff
@@ -47,6 +47,9 @@ typedef char PlayerFlags[ PLAYER_FLAGS_MAX ];
 #define player_isTut()     player_isFlag(PLAYER_TUTORIAL)
 
+/* dt_default grabber helper */
+#define player_dt_default() (player.p != NULL && player.p->ship != NULL ? player.p->ship->dt_default : 1)
```
I don't know if it really matters in this case, but I'd probably be inclined to use a function instead of a macro when possible. A modern C compiler can inline the function to have the same effect.

onpon4 replied:
OK, done. It's a function now.

onpon4 commented:
I've added an include for assert.h  and changed player_dt_default to a function.

I'm not sure about switching up the way dockpilot and dockslot are stored, it's not going to improve efficiency any (and in fact can make it worse because it adds more looping). That it involves passing pointers around is true enough but failure to clean up the other values properly can cause bugs too (just bugs that would be less obvious like choosing the wrong ship as a dockpilot and then trying to dock on a slot that it doesn't belong in).

onpon4 commented:
Now, if every single ship that could possibly have the destroyed pilot as a dockslot will always be in the escort list, that can be used to reduce looping a bit. That looks rather flimsy to me, though, and I don't think that's really worth it when the pilots have to be looped through all the time anyway.

onpon4 commented:
There we go, switched it up to use ints instead of pointers.

ids1024 commented:
Do you see my point about just removing the loop from `pilot_destroy`, or am I missing something? All that code does is setting `dockpilot` and `dockslot` to `0` and `-1`. Without that loop, those variables will be set the same way the next time `pilot_getDockSlot()` is called.

Otherwise, the changes here look good. :+1:

onpon4 commented:
Ahhhhhhh, I see what you're saying. That does make sense come to think of it. Lemme do that...

onpon4 commented:
There we go, I've pushed that through. Seems to work.
